go/consensus/tendermint/apps/governance/transactions.go

```diff
@@ -0,0 +1,288 @@
+		CreatedAt: epoch,
+		Deposit:   params.MinProposalDeposit,
+		State:     governance.StateActive,
+		Submitter: submitterAddr,
+	}
+	if err := state.SetActiveProposal(ctx, proposal); err != nil {
+		ctx.Logger().Error("governance: failed to set active proposal",
+			"err", err,
+		)
+		return fmt.Errorf("governance: failed to set active proposal: %w", err)
+	}
+
+	// Emit events.
+	// Proposal submitted.
+	evt := &governance.ProposalSubmittedEvent{
+		ID:        proposal.ID,
+		Submitter: proposal.Submitter,
+	}
+	ctx.EmitEvent(api.NewEventBuilder(app.Name()).Attribute(KeyProposalSubmitted, cbor.Marshal(evt)))
+
+	return nil
+}
+
+func (app *governanceApplication) castVote(
+	ctx *api.Context,
+	state *governanceState.MutableState,
+	proposalVote *governance.ProposalVote,
+) error {
+	params, err := state.ConsensusParameters(ctx)
+	if err != nil {
+		return fmt.Errorf("governance: failed to fetch consensus parameters: %w", err)
+	}
+
+	submitterAddr := stakingAPI.NewAddress(ctx.TxSigner())
+	if !submitterAddr.IsValid() {
+		return stakingAPI.ErrForbidden
+	}
+
+	// Query signer entity descriptor.
+	registryState := registryState.NewMutableState(ctx.State())
+	submitterEntity, err := registryState.Entity(ctx, ctx.TxSigner())
+	switch err {
+	case nil:
+	case registryAPI.ErrNoSuchEntity:
+		return governance.ErrNotEligible
+	default:
+		return fmt.Errorf("governance: failed to query entity: %w", err)
+	}
+	schedulerState := schedulerState.NewMutableState(ctx.State())
+	currentValidators, err := schedulerState.CurrentValidators(ctx)
+	if err != nil {
+		return fmt.Errorf("governance: failed to query current validators: %w", err)
+	}
+	// Submitter is eligible if any of its nodes is part of the current validator committee.
+	var eligible bool
+	for _, nID := range submitterEntity.Nodes {
+		var node *node.Node
+		node, err = registryState.Node(ctx, nID)
+		if err != nil {
+			return fmt.Errorf("governance: failed to query entity node: %w", err)
+		}
+		if _, ok := currentValidators[node.Consensus.ID]; ok {
+			eligible = true
+			break
+		}
+	}
+	if !eligible {
+		ctx.Logger().Error("governance: submitter not eligible to vote",
+			"submitter", ctx.TxSigner(),
+		)
+		return governance.ErrNotEligible
+	}
+
+	// Load proposal.
+	proposal, err := state.Proposal(ctx, proposalVote.ID)
+	switch err {
+	case nil:
+	case governance.ErrNoSuchProposal:
+		ctx.Logger().Error("governance: vote for a missing proposal",
+			"proposal_id", proposalVote.ID,
+		)
+		return governance.ErrNoSuchProposal
+	default:
+		ctx.Logger().Error("governance: error loading proposal",
+			"err", err,
+			"proposal_id", proposalVote.ID,
+		)
+	}
+	// Ensure proposal is active.
+	if proposal.State != governance.StateActive {
+		ctx.Logger().Error("governance: vote for a non-active proposal",
+			"proposal_id", proposalVote.ID,
+			"state", proposal.State,
+			"proposal", proposal,
+			"vote", proposalVote,
+		)
+		return governance.ErrVotingIsClosed
+	}
+
+	if ctx.IsCheckOnly() {
```

Comment:
This seems like it is doing a lot in CheckTx, maybe avoid it. If there's nothing stateless that can be checked in CheckTx, feel free to skip those checks.

Reply:
ok makes sense yeah. Fixed.